### PR TITLE
simplify pipelining of resource ids

### PR DESCRIPTION
**Pull request by cadecairos**
#148 

This patch only looks for keys on the action data that we actually care about (id, projectId, pageId) and processes them.

I don't know what kind of perf improvements we'll see, but I bet it's a start.

### CODE REVIEW

File: services/api/lib/bulk.js [hunk 1]
```diff
@@ -152,21 +150,39 @@ exports.register = function(server, options, done) {
...
-      // if the key's value isn't a reach string, return valid
-      if ( !reachRegex.test(action.data[key]) ) {
-        return processResult;
+  function reachForData(action, actionIndex, txResults) {
+    var key;
+
+    for (var i = 0; i < pipelineKeys.length; i++) {
+      if (action.data[pipelineKeys[i]]) {
+        key = pipelineKeys[i];
+        continue;
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: why is this here?

File: services/api/lib/bulk.js [hunk 2]
```diff
@@ -116,7 +116,6 @@ exports.register = function(server, options, done) {
     return function replace(match, reachIdx, reachString) {
       reachIdx = +reachIdx;
       var value;
-
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: whitespace change?

File: services/api/lib/bulk.js (hunk 1 above)
Comment (cadecairos) [inline services/api/lib/bulk.js]: if we match, no need to keep iterating.

File: services/api/lib/bulk.js [hunk 3]
```diff
@@ -129,7 +128,6 @@ exports.register = function(server, options, done) {
         };
         return;
       }
-
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: whitespace change?

File: services/api/lib/bulk.js [hunk 4]
```diff
@@ -152,21 +152,39 @@ exports.register = function(server, options, done) {
...
-      return processResult;
-    };
+    if (!key) {
+      return true;
+    }
+
+    // if the key's value isn't a reach string, return valid
+    if ( !reachRegex.test(action.data[key]) ) {
+      return true;
+    }
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: could this be reduced to something like:
```
if ( !key || !reachRegex.test(action.data[key])) { return true; }
```

File: services/api/lib/bulk.js [hunk 5]
```diff
@@ -152,21 +152,39 @@ exports.register = function(server, options, done) {
...
-        getReplaceFunc(processResult, key, actionIndex, txResults)
-      );
-      return processResult;
-    };
+    if (!key) {
+      return true;
+    }
+
+    // if the key's value isn't a reach string, return valid
+    if ( !reachRegex.test(action.data[key]) ) {
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: maybe put this in a var with a more meaningful name? hard to read.

File: services/api/lib/bulk.js [hunk 6]
```diff
@@ -152,21 +152,39 @@ exports.register = function(server, options, done) {
...
+  ];
+
   // check if a key on the action object should be resolved to a value
   // returned by a previous action in the transaction
-  function generateForEachCallback(processResult, action, actionIndex, txResults) {
-    return function everyCallback(key) {
-      // if the key's value isn't a reach string, return valid
-      if ( !reachRegex.test(action.data[key]) ) {
-        return processResult;
+  function reachForData(action, actionIndex, txResults) {
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: this is still a very large function- do we want to try to simplify it into smaller util functions? i think the iteration (lines 166-171) could be one, also the replace function in lines 182-187.

this way this function is really just 2 functions and some checks. much easier to reason about i think (function names help explain what is going on)

File: services/api/lib/bulk.js [hunk 7]
```diff
@@ -152,21 +152,39 @@ exports.register = function(server, options, done) {
...
+    if (!key) {
+      return true;
+    }
+
+    // if the key's value isn't a reach string, return valid
+    if ( !reachRegex.test(action.data[key]) ) {
+      return true;
+    }
+
+    var replaceResult = {};
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: if we are not refactoring this out into a separate function i think all variable declarations should be at the top of the file. not sure what everyone else's feelings are on that; my reasoning: hoisting.

File: services/api/lib/bulk.js [hunk 8]
```diff
@@ -152,21 +152,40 @@ exports.register = function(server, options, done) {
...
-        getReplaceFunc(processResult, key, actionIndex, txResults)
-      );
-      return processResult;
-    };
+    // if the key's value isn't a reach string, return valid
+    if ( !key || !isReachString(data[key]) ) {
+      return true;
+    }
+
+    var replaceResult = {};
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: how about making this a separate function as well?

**Comment by cadecairos** [comment]
@ashleygwilliams okay updated again!

File: services/api/lib/postgre.js [hunk 9]
```diff
@@ -314,24 +314,16 @@ module.exports = function (pg) {
...
-
-            // Check every key of the action's data to see if it should be replaced with
-            // the result of a previous action.
-            var dataKeys = Object.keys(action.data);
-            server.methods.newrelic.createTracer('pipelining action data', dataKeys.forEach);
-            dataKeys.forEach(
-              server.methods.bulk.generateForEachCallback(processResult, action, actionIndex, results)
-            );
+            // Check the action's data to see if it ids should be replaced with the results of a previous action.
+            server.methods.newrelic.createTracer('pipelining action data', server.methods.bulk.reachForData);
```
Comment (ashleygwilliams) [inline services/api/lib/postgre.js]: do we want this new relic stuff in here? seems out of scope for this PR or should at least be it's own commit.

Comment (cadecairos) [inline services/api/lib/postgre.js]: it was already there

Comment (cadecairos) [inline services/api/lib/postgre.js]: @ashleygwilliams ^

File: services/api/lib/bulk.js [hunk 10]
```diff
@@ -152,21 +152,44 @@ exports.register = function(server, options, done) {
...
+        return pipelineKeys[i];
+      }
+    }
+  }
+
+  function isReachString(value) {
+    return reachRegex.test(value);
+  }
+
+  function replaceValue(data, key, actionIndex, txResults) {
```
Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: this function makes me uneasy-- it takes in certain parameters, does a replace, and returns a side effect(?) of a function that is called during a replace on a string that... doesn't get used again?  maybe i am misunderstanding but i think that we should consider rewriting it so that it better expresses the purpose of the code.

Comment (cadecairos) [inline services/api/lib/bulk.js]: What doesn't get used again? the side effect (replaceResult) lets us provide a meaningful error if something goes wrong.

Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: `data[key] = data[key].replace...`

from what i can tell the "new" `data[key]` does not get used?

Comment (cadecairos) [inline services/api/lib/bulk.js]: it's getting used when inserting/updating/deleting?

Comment (ashleygwilliams) [inline services/api/lib/bulk.js]: it's not returned from the function... is it global?? (i assume not since it is passed into the func)

Comment (cadecairos) [inline services/api/lib/bulk.js]: it's an object passed in on line 184.

since it's pass by reference, it gets set to the right value for when we need it.

**Comment by ashleygwilliams** [comment]
this code really improved- we should file an issue to refactor the rest of the code as well.

also, i think in the long run it would be best to change `postgres.js` and `bulk.js` into directories with an `index` file and a `lib` directory where the `lib` held smaller files of grouped utility functions. the file sizes are still extremely long, and the grouping would make the intention of the code flow with the utility functions even clearer.

i'm r+ for this with an eye towards more refactoring in the near future.

**Comment by ashleygwilliams** [comment]
oh also-- squash!
